// rubic-sdk/lib/features/cross-chain/providers/rango-trade-provider/rango-cross-chain-trade.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RangoCrossChainTrade = void 0;
var bignumber_js_1 = __importDefault(require("bignumber.js"));
var core_1 = require("../../../../core");
var features_1 = require("../../..");
var injector_1 = require("../../../../core/sdk/injector");
var lib_1 = require("rango-sdk-basic/lib");
var common_1 = require("../../../../common");
var not_whitelisted_provider_error_1 = require("../../../../common/errors/swap/not-whitelisted-provider.error");
var empty_address_1 = require("../../../../core/blockchain/constants/empty-address");
var cross_chain_trade_1 = require("../common/cross-chain-trade");
var get_from_without_fee_1 = require("../../utils/get-from-without-fee");
var contract_address_1 = require("./constants/contract-address");
var common_cross_chain_abi_1 = require("../common/constants/common-cross-chain-abi");
var rango_api_key_1 = require("./constants/rango-api-key");
var rango_blockchain_name_1 = require("./constants/rango-blockchain-name");
var RangoCrossChainTrade = /** @class */ (function (_super) {
    __extends(RangoCrossChainTrade, _super);
    function RangoCrossChainTrade(crossChainTrade, rangoClientRef, providerAddress) {
        var _this = _super.call(this, providerAddress) || this;
        _this.type = features_1.CROSS_CHAIN_TRADE_TYPE.RANGO;
        _this.from = crossChainTrade.from;
        _this.to = crossChainTrade.to;
        _this.toTokenAmountMin = crossChainTrade.toTokenAmountMin;
        _this.feeInfo = crossChainTrade.feeInfo;
        _this.slippageTolerance = crossChainTrade.slippageTolerance;
        _this.bridgeType = crossChainTrade.bridgeType;
        _this.itType = crossChainTrade.itType;
        _this.priceImpact = crossChainTrade.priceImpact;
        _this.cryptoFeeToken = crossChainTrade.cryptoFeeToken;
        _this.gasData = crossChainTrade.gasData;
        _this.fromWeb3Public = injector_1.Injector.web3PublicService.getWeb3Public(_this.from.blockchain);
        _this.rangoClientRef = rangoClientRef;
        return _this;
    }
    /** @internal */
    RangoCrossChainTrade.getGasData = function (from, to) {
        return __awaiter(this, void 0, void 0, function () {
            var fromBlockchain, web3Public, walletAddress, _a, contractAddress, contractAbi, methodName, methodArguments, value, _b, gasLimit, gasPrice, _c, _d, _e, _f, increasedGasLimit, _err_1;
            return __generator(this, function (_g) {
                switch (_g.label) {
                    case 0:
                        fromBlockchain = from.blockchain;
                        web3Public = injector_1.Injector.web3PublicService.getWeb3Public(fromBlockchain);
                        walletAddress = injector_1.Injector.web3Private.address;
                        if (!walletAddress) {
                            return [2 /*return*/, null];
                        }
                        _g.label = 1;
                    case 1:
                        _g.trys.push([1, 5, , 6]);
                        return [4 /*yield*/, new RangoCrossChainTrade({
                                from: from,
                                to: to,
                                toTokenAmountMin: new bignumber_js_1.default(0),
                                slippageTolerance: 4,
                                cryptoFeeToken: {},
                                feeInfo: {
                                    cryptoFee: null,
                                    fixedFee: null,
                                    platformFee: null
                                },
                                itType: {
                                    from: undefined,
                                    to: undefined
                                },
                                bridgeType: undefined,
                                priceImpact: null,
                                gasData: {
                                    gasLimit: new bignumber_js_1.default(0),
                                    gasPrice: new bignumber_js_1.default(0)
                                }
                            }, new lib_1.RangoClient(rango_api_key_1.RANGO_API_KEY), empty_address_1.EMPTY_ADDRESS).getContractParams()];
                    case 2:
                        _a = _g.sent(), contractAddress = _a.contractAddress, contractAbi = _a.contractAbi, methodName = _a.methodName, methodArguments = _a.methodArguments, value = _a.value;
                        _d = (_c = Promise).all;
                        _e = [web3Public.getEstimatedGas(contractAbi, contractAddress, methodName, methodArguments, walletAddress, value)];
                        _f = bignumber_js_1.default.bind;
                        return [4 /*yield*/, injector_1.Injector.gasPriceApi.getGasPrice(fromBlockchain)];
                    case 3: return [4 /*yield*/, _d.apply(_c, [_e.concat([
                                new (_f.apply(bignumber_js_1.default, [void 0, _g.sent()]))()
                            ])])];
                    case 4:
                        _b = _g.sent(), gasLimit = _b[0], gasPrice = _b[1];
                        if (!(gasLimit === null || gasLimit === void 0 ? void 0 : gasLimit.isFinite())) {
                            return [2 /*return*/, null];
                        }
                        increasedGasLimit = core_1.Web3Pure.calculateGasMargin(gasLimit, 1.2);
                        return [2 /*return*/, {
                                gasLimit: increasedGasLimit,
                                gasPrice: gasPrice
                            }];
                    case 5:
                        _err_1 = _g.sent();
                        return [2 /*return*/, null];
                    case 6: return [2 /*return*/];
                }
            });
        });
    };
    Object.defineProperty(RangoCrossChainTrade.prototype, "fromBlockchain", {
        get: function () {
            return this.from.blockchain;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RangoCrossChainTrade.prototype, "fromContractAddress", {
        get: function () {
            return contract_address_1.RANGO_CONTRACT_ADDRESSES[this.fromBlockchain].rubicRouter;
        },
        enumerable: false,
        configurable: true
    });
    RangoCrossChainTrade.prototype.swap = function (options) {
        var _a, _b;
        if (options === void 0) { options = {}; }
        return __awaiter(this, void 0, void 0, function () {
            var onConfirm, _c, contractAddress, contractAbi, methodName, methodArguments, value, transactionHash, onTransactionHash, err_1;
            return __generator(this, function (_d) {
                switch (_d.label) {
                    case 0:
                        if (options === null || options === void 0 ? void 0 : options.receiverAddress) {
                            throw new common_1.UnsupportedReceiverAddressError();
                        }
                        return [4 /*yield*/, this.checkTradeErrors()];
                    case 1:
                        _d.sent();
                        return [4 /*yield*/, this.checkAllowanceAndApprove(options)];
                    case 2:
                        _d.sent();
                        onConfirm = options.onConfirm;
                        return [4 /*yield*/, this.getContractParams()];
                    case 3:
                        _c = _d.sent(), contractAddress = _c.contractAddress, contractAbi = _c.contractAbi, methodName = _c.methodName, methodArguments = _c.methodArguments, value = _c.value;
                        onTransactionHash = function (hash) {
                            if (onConfirm) {
                                onConfirm(hash);
                            }
                            transactionHash = hash;
                        };
                        _d.label = 4;
                    case 4:
                        _d.trys.push([4, 6, , 7]);
                        return [4 /*yield*/, injector_1.Injector.web3Private.tryExecuteContractMethod(contractAddress, contractAbi, methodName, methodArguments, {
                                value: value,
                                onTransactionHash: onTransactionHash,
                                gas: (_a = this.gasData) === null || _a === void 0 ? void 0 : _a.gasLimit,
                                gasPrice: (_b = this.gasData) === null || _b === void 0 ? void 0 : _b.gasPrice
                            })];
                    case 5:
                        _d.sent();
                        return [2 /*return*/, transactionHash];
                    case 6:
                        err_1 = _d.sent();
                        if (err_1 instanceof common_1.FailedToCheckForTransactionReceiptError) {
                            return [2 /*return*/, transactionHash];
                        }
                        throw err_1;
                    case 7: return [2 /*return*/];
                }
            });
        });
    };
    RangoCrossChainTrade.prototype.getContractParams = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _a, txData, providerValue, txTo, routerCallParams, methodArguments, value;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, this.refetchTxData()];
                    case 1:
                        _a = _b.sent(), txData = _a.txData, providerValue = _a.value, txTo = _a.txTo;
                        return [4 /*yield*/, this.checkProviderIsWhitelisted(txTo)];
                    case 2:
                        _b.sent();
                        routerCallParams = [
                            this.from.address,
                            this.from.stringWeiAmount,
                            core_1.BlockchainsInfo.getBlockchainByName(this.to.blockchain).id,
                            this.to.address,
                            core_1.Web3Pure.toWei(this.toTokenAmountMin, this.to.decimals),
                            this.walletAddress,
                            this.providerAddress,
                            txTo
                        ];
                        methodArguments = [
                            "".concat(this.type.toLowerCase(), ":").concat(this.bridgeType),
                            routerCallParams
                        ];
                        if (!this.from.isNative) {
                            methodArguments.push(txTo);
                        }
                        methodArguments.push(txData);
                        value = this.getSwapValue(providerValue);
                        return [2 /*return*/, {
                                contractAddress: this.fromContractAddress,
                                contractAbi: common_cross_chain_abi_1.commonCrossChainAbi,
                                methodName: this.methodName,
                                methodArguments: methodArguments,
                                value: value
                            }];
                }
            });
        });
    };
    RangoCrossChainTrade.prototype.refetchTxData = function () {
        return __awaiter(this, void 0, void 0, function () {
            var amountWithoutFee, response;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        amountWithoutFee = (0, get_from_without_fee_1.getFromWithoutFee)(this.from, this.feeInfo).stringWeiAmount;
                        return [4 /*yield*/, this.rangoClientRef.swap({
                                from: {
                                    blockchain: rango_blockchain_name_1.RANGO_BLOCKCHAIN_NAME[this.from.blockchain],
                                    symbol: this.from.symbol,
                                    address: this.from.isNative ? null : this.from.address
                                },
                                to: {
                                    blockchain: rango_blockchain_name_1.RANGO_BLOCKCHAIN_NAME[this.to.blockchain],
                                    symbol: this.to.symbol,
                                    address: this.to.isNative ? null : this.to.address
                                },
                                amount: amountWithoutFee,
                                disableEstimate: false,
                                slippage: String(this.slippageTolerance * 100),
                                fromAddress: this.walletAddress,
                                toAddress: this.walletAddress,
                                referrerAddress: null,
                                referrerFee: null
                            })];
                    case 1:
                        response = _a.sent();
                        this.requestId = response.requestId;
                        return [2 /*return*/, response.tx];
                }
            });
        });
    };
    RangoCrossChainTrade.prototype.checkProviderIsWhitelisted = function (providerRouter) {
        return __awaiter(this, void 0, void 0, function () {
            var whitelistedContracts;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, injector_1.Injector.web3PublicService
                            .getWeb3Public(this.from.blockchain)
                            .callContractMethod(this.fromContractAddress, common_cross_chain_abi_1.commonCrossChainAbi, 'getAvailableRouters')];
                    case 1:
                        whitelistedContracts = _a.sent();
                        if (!whitelistedContracts.find(function (whitelistedContract) {
                            return (0, common_1.compareAddresses)(whitelistedContract, providerRouter);
                        })) {
                            throw new not_whitelisted_provider_error_1.NotWhitelistedProviderError(providerRouter);
                        }
                        return [2 /*return*/];
                }
            });
        });
    };
    RangoCrossChainTrade.prototype.getTradeAmountRatio = function (fromUsd) {
        var usdCryptoFee = this.cryptoFeeToken.price.multipliedBy(this.cryptoFeeToken.tokenAmount);
        return fromUsd.plus(usdCryptoFee.isNaN() ? 0 : usdCryptoFee).dividedBy(this.to.tokenAmount);
    };
    return RangoCrossChainTrade;
}(cross_chain_trade_1.CrossChainTrade));
exports.RangoCrossChainTrade = RangoCrossChainTrade;
