// rubic-sdk/lib/features/instant-trades/dexes/common/rango/constants/rango-providers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.rangoProviders = void 0;
var trade_type_1 = require("../../../../models/trade-type");
exports.rangoProviders = {
    SpookySwap: trade_type_1.TRADE_TYPE.SPOOKY_SWAP,
    OneInchArbitrum: trade_type_1.TRADE_TYPE.ONE_INCH,
    OneInchBsc: trade_type_1.TRADE_TYPE.ONE_INCH,
    OneInchEth: trade_type_1.TRADE_TYPE.ONE_INCH,
    OneInchGnosis: trade_type_1.TRADE_TYPE.ONE_INCH,
    OneInchOptimism: trade_type_1.TRADE_TYPE.ONE_INCH,
    OneInchPolygon: trade_type_1.TRADE_TYPE.ONE_INCH,
    TrisolarisSwap: trade_type_1.TRADE_TYPE.TRISOLARIS,
    AuroraSwap: trade_type_1.TRADE_TYPE.AURORA_SWAP,
    BeamSwap: trade_type_1.TRADE_TYPE.BEAM_SWAP,
    CronaSwap: trade_type_1.TRADE_TYPE.CRONA_SWAP,
    MMFinance: trade_type_1.TRADE_TYPE.MM_FINANCE,
    VVSFinance: trade_type_1.TRADE_TYPE.VVS_FINANCE,
    Jupiter: trade_type_1.TRADE_TYPE.JUPITER,
    OpenOceanFantom: trade_type_1.TRADE_TYPE.OPEN_OCEAN,
    PancakeSwapBsc: trade_type_1.TRADE_TYPE.PANCAKE_SWAP,
    PangolinSwap: trade_type_1.TRADE_TYPE.PANGOLIN,
    'ParaSwap Avalanche': trade_type_1.TRADE_TYPE.PARA_SWAP,
    'ParaSwap Polygon': trade_type_1.TRADE_TYPE.PARA_SWAP,
    QuickSwap: trade_type_1.TRADE_TYPE.QUICK_SWAP,
    SolarbeamSwap: trade_type_1.TRADE_TYPE.SOLAR_BEAM,
    StellaSwap: trade_type_1.TRADE_TYPE.STELLA_SWAP,
    SushiArbitrum: trade_type_1.TRADE_TYPE.SUSHI_SWAP,
    EthUniSwapv2: trade_type_1.TRADE_TYPE.UNISWAP_V2,
    ViperSwap: trade_type_1.TRADE_TYPE.VIPER_SWAP,
    VoltageSwap: trade_type_1.TRADE_TYPE.VOLTAGE_SWAP
};
