// @renproject/chains-ethereum/module/goerli.js
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import { EthereumBaseChain } from "./base";
import { goerliConfigMap } from "./ethereum";
import { resolveEVMNetworkConfig } from "./utils/generic";
export class Goerli extends EthereumBaseChain {
    constructor(_a) {
        var { network } = _a, params = __rest(_a, ["network"]);
        super(Object.assign(Object.assign({}, params), { network: resolveEVMNetworkConfig(goerliConfigMap, network) }));
        this.configMap = goerliConfigMap;
        this.assets = Goerli.assets;
    }
}
// Static members.
Goerli.chain = "Goerli";
Goerli.configMap = goerliConfigMap;
Goerli.assets = {
    gETH: "gETH",
    REN: "REN_Goerli",
    DAI: "DAI_Goerli",
    USDC: "USDC_Goerli",
    USDT: "USDT_Goerli",
    // Aliases
    ETH: "gETH",
    ETH_Goerli: "gETH",
    REN_Goerli: "REN_Goerli",
    DAI_Goerli: "DAI_Goerli",
    USDC_Goerli: "USDC_Goerli",
    USDT_Goerli: "USDT_Goerli",
};
