// @renproject/chains-ethereum/goerli.js
"use strict";
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Goerli = void 0;
const base_1 = require("./base");
const ethereum_1 = require("./ethereum");
const generic_1 = require("./utils/generic");
class Goerli extends base_1.EthereumBaseChain {
    constructor(_a) {
        var { network } = _a, params = __rest(_a, ["network"]);
        super(Object.assign(Object.assign({}, params), { network: (0, generic_1.resolveEVMNetworkConfig)(ethereum_1.goerliConfigMap, network) }));
        this.configMap = ethereum_1.goerliConfigMap;
        this.assets = Goerli.assets;
    }
}
exports.Goerli = Goerli;
// Static members.
Goerli.chain = "Goerli";
Goerli.configMap = ethereum_1.goerliConfigMap;
Goerli.assets = {
    gETH: "gETH",
    REN: "REN_Goerli",
    DAI: "DAI_Goerli",
    USDC: "USDC_Goerli",
    USDT: "USDT_Goerli",
    // Aliases
    ETH: "gETH",
    ETH_Goerli: "gETH",
    REN_Goerli: "REN_Goerli",
    DAI_Goerli: "DAI_Goerli",
    USDC_Goerli: "USDC_Goerli",
    USDT_Goerli: "USDT_Goerli",
};
