// @renproject/chains-ethereum/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Polygon = exports.Optimism = exports.Moonbeam = exports.Kava = exports.Goerli = exports.Fantom = exports.Ethereum = exports.Catalog = exports.BinanceSmartChain = exports.Avalanche = exports.Arbitrum = exports.resolveEVMNetworkConfig = exports.isEVMNetworkConfig = exports.resolveRpcEndpoints = exports.EthereumBaseChain = exports.EVM_ERROR = exports.EVMParam = void 0;
__exportStar(require("./utils/types"), exports);
var evmParams_1 = require("./utils/payloads/evmParams");
Object.defineProperty(exports, "EVMParam", { enumerable: true, get: function () { return evmParams_1.EVMParam; } });
var errors_1 = require("./utils/errors");
Object.defineProperty(exports, "EVM_ERROR", { enumerable: true, get: function () { return errors_1.EVM_ERROR; } });
var base_1 = require("./base");
Object.defineProperty(exports, "EthereumBaseChain", { enumerable: true, get: function () { return base_1.EthereumBaseChain; } });
var generic_1 = require("./utils/generic");
Object.defineProperty(exports, "resolveRpcEndpoints", { enumerable: true, get: function () { return generic_1.resolveRpcEndpoints; } });
Object.defineProperty(exports, "isEVMNetworkConfig", { enumerable: true, get: function () { return generic_1.isEVMNetworkConfig; } });
Object.defineProperty(exports, "resolveEVMNetworkConfig", { enumerable: true, get: function () { return generic_1.resolveEVMNetworkConfig; } });
// Chains
var arbitrum_1 = require("./arbitrum");
Object.defineProperty(exports, "Arbitrum", { enumerable: true, get: function () { return arbitrum_1.Arbitrum; } });
var avalanche_1 = require("./avalanche");
Object.defineProperty(exports, "Avalanche", { enumerable: true, get: function () { return avalanche_1.Avalanche; } });
var bsc_1 = require("./bsc");
Object.defineProperty(exports, "BinanceSmartChain", { enumerable: true, get: function () { return bsc_1.BinanceSmartChain; } });
var catalog_1 = require("./catalog");
Object.defineProperty(exports, "Catalog", { enumerable: true, get: function () { return catalog_1.Catalog; } });
var ethereum_1 = require("./ethereum");
Object.defineProperty(exports, "Ethereum", { enumerable: true, get: function () { return ethereum_1.Ethereum; } });
var fantom_1 = require("./fantom");
Object.defineProperty(exports, "Fantom", { enumerable: true, get: function () { return fantom_1.Fantom; } });
var goerli_1 = require("./goerli");
Object.defineProperty(exports, "Goerli", { enumerable: true, get: function () { return goerli_1.Goerli; } });
var kava_1 = require("./kava");
Object.defineProperty(exports, "Kava", { enumerable: true, get: function () { return kava_1.Kava; } });
var moonbeam_1 = require("./moonbeam");
Object.defineProperty(exports, "Moonbeam", { enumerable: true, get: function () { return moonbeam_1.Moonbeam; } });
var optimism_1 = require("./optimism");
Object.defineProperty(exports, "Optimism", { enumerable: true, get: function () { return optimism_1.Optimism; } });
var polygon_1 = require("./polygon");
Object.defineProperty(exports, "Polygon", { enumerable: true, get: function () { return polygon_1.Polygon; } });
